### PR TITLE
Introduce basic transactions API

### PR DESCRIPTION
**Pull request by kandrio**
Everything is summarized in our commit messages and on the following issue of our team:
- https://github.com/Tribler/tribler/issues/7293

### CODE REVIEW

File: detoks/src/main/java/nl/tudelft/trustchain/detoks/DeToksCommunity.kt [hunk 1]
```diff
@@ -2,13 +2,14 @@ package nl.tudelft.trustchain.detoks
...
-import nl.tudelft.ipv8.Community
 import nl.tudelft.ipv8.Overlay
 import nl.tudelft.ipv8.Peer
 import nl.tudelft.ipv8.messaging.Packet
 
 
-class DeToksCommunity(private val context: Context) : Community() {
+class DeToksCommunity(
+    private val context: Context
+    ) : TransactionEngine("c86a7db45eb3563ae047639817baec4db2bc7c25") {
```
Comment (OrestisKan) [inline detoks/src/main/java/nl/tudelft/trustchain/detoks/DeToksCommunity.kt:12]: I understand that serviceId was hardcoded in the code but it seems to me that it should not be, maybe use some environment file or something to hide it?

File: detoks/src/main/java/nl/tudelft/trustchain/detoks/LightweightTxApp.kt [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    private val scope = CoroutineScope(Dispatchers.Default)
+
+    fun run() {
+        startIpv8()
+    }
+
+    private fun createCommunity(): OverlayConfiguration<TransactionEngine> {
+        val randomWalk = RandomWalk.Factory(timeout = 3.0, peers = 20)
+        return OverlayConfiguration(
+            TransactionEngine.Factory("c86a7db45eb3563ae047639817baec4db2bc7c25"),
```
Comment (OrestisKan) [inline detoks/src/main/java/nl/tudelft/trustchain/detoks/LightweightTxApp.kt:31]: Again hardcoded

File: detoks/src/main/java/nl/tudelft/trustchain/detoks/LightweightTxApp.kt [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+        val ipv8 = IPv8(endpoint, config, myPeer)
+        ipv8.start()
+
+        val numberOfTransactions = 1
+        scope.launch {
+            while (true) {
+                val overlay = ipv8.overlays.values.toList()[0]
+                printPeersInfo(overlay)
+                println("===")
```
Comment (OrestisKan) [inline detoks/src/main/java/nl/tudelft/trustchain/detoks/LightweightTxApp.kt:59]: remove

File: detoks/src/main/java/nl/tudelft/trustchain/detoks/LightweightTxApp.kt [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        val driver: SqlDriver = JdbcSqliteDriver(JdbcSqliteDriver.IN_MEMORY)
+        Database.Schema.create(driver)
+        val store = TrustChainSQLiteStore(Database(driver))
+
+        val ipv8 = IPv8(endpoint, config, myPeer)
+        ipv8.start()
+
+        val numberOfTransactions = 1
+        scope.launch {
+            while (true) {
```
Comment (OrestisKan) [inline detoks/src/main/java/nl/tudelft/trustchain/detoks/LightweightTxApp.kt:56]: Bad practice

File: detoks/src/main/java/nl/tudelft/trustchain/detoks/TransactionEngine.kt [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+    init {
+        messageHandlers[msgIdFixer(MessageId.HALF_BLOCK)] = ::onHalfBlockPacket
+        messageHandlers[msgIdFixer(MessageId.HALF_BLOCK_BROADCAST)] = ::onHalfBlockBroadcastPacket
+        messageHandlers[msgIdFixer(MessageId.HALF_BLOCK_ENCRYPTED)] = ::onHalfBlockPacket
+        messageHandlers[msgIdFixer(MessageId.HALF_BLOCK_BROADCAST_ENCRYPTED)] = ::onHalfBlockBroadcastPacket
+    }
+
+    fun sendTransaction(blockBuilder: BlockBuilder, peer: Peer?, encrypt: Boolean = false) {
+        println("sending block...")
```
Comment (OrestisKan) [inline detoks/src/main/java/nl/tudelft/trustchain/detoks/TransactionEngine.kt:34]: Debug print ? Use logger if needed

File: detoks/src/main/java/nl/tudelft/trustchain/detoks/TransactionEngine.kt [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+        val handler = messageHandlers[msgId]
+
+        if (handler != null) {
+            try {
+                handler(packet)
+            } catch (e: Exception) {
+                e.printStackTrace()
+            }
+        } else {
+            println("Received unknown message $msgId from $sourceAddress")
```
Comment (OrestisKan) [inline detoks/src/main/java/nl/tudelft/trustchain/detoks/TransactionEngine.kt:103]: Use logger

File: detoks/src/main/java/nl/tudelft/trustchain/detoks/TransactionEngine.kt [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        }
+    }
+
+    private fun msgIdFixer(msgid: Int): Int{
+        @Suppress("DEPRECATION")
```
Comment (OrestisKan) [inline detoks/src/main/java/nl/tudelft/trustchain/detoks/TransactionEngine.kt:108]: Not a good practice, there should be some other approach for the same action

**Comment by kandrio** [comment]
Thank you very much for the comments. We will address them as soon as
possible.

Στις Τρί, 28 Μαρ 2023, 14:15 ο χρήστης Orestis Kanaris <
***@***.***> έγραψε:

File: detoks/src/main/java/nl/tudelft/trustchain/detoks/LightweightTxApp.kt (hunk 3 above)
Comment (kandrio) [reply detoks/src/main/java/nl/tudelft/trustchain/detoks/LightweightTxApp.kt:59]: Thanks for the comment. Could you be a bit more specific here? Are you referring to the entire `while (true)` statement? Or to the `printPeersInfo(overlay)` part?

File: detoks/src/main/java/nl/tudelft/trustchain/detoks/TransactionEngine.kt (hunk 7 above)
Comment (kandrio) [reply detoks/src/main/java/nl/tudelft/trustchain/detoks/TransactionEngine.kt:108]: I agree. The problem was that we used `MessageId` flags (e.g: `HALF_BLOCK`, `HALF_BLOCK_ENCRYPTED` etc...) that were larger than a `Ubyte`. I will use smaller values as flags and I will remove the `msgIdFixer`.

File: detoks/src/main/java/nl/tudelft/trustchain/detoks/TransactionEngine.kt (hunk 5 above)
Comment (kandrio) [reply detoks/src/main/java/nl/tudelft/trustchain/detoks/TransactionEngine.kt:34]: I agree. I will use a logger from the `mu.KotlinLogging` library

**Comment by kandrio** [comment]
@OrestisKan We updated our PR and addressed all of your comments. We decided to drop the `LightweightTxApp` from the PR. We'll keep it internally for testing and won't push it as a "production-ready" code. 

Feel free to take a look again at our branch. We're to clarify any questions you may have

**Comment by synctext** [comment]
@OrestisKan is a bit overloaded, as nobody made the ```Week 6``` deadline for Prototype I. Review will come asap.

**Review by OrestisKan — CHANGES_REQUESTED** [review]
Code seems alright, but I'm not sure how to work the app, maybe you can guide me ? 
Also fix the merge conflicts please.

**Review by OrestisKan — APPROVED** [review]
Seems alright!

**Pull request merged**